libraries/lib-utility/BufferedStreamReader.h

```diff
@@ -0,0 +1,111 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  BufferedStreamReader.h
+
+  Dmitry Vedenko
+
+**********************************************************************/
+
+#pragma once
+
+#include <vector>
+#include <cstdint>
+#include <type_traits>
+
+/*!
+ * \brief A facade-like class, that implements buffered reading from the underlying data stream.
+ *
+ * BufferedStreamReader provides optimal read performance for the built-in types of up to 8 bytes long.
+ */
+class UTILITY_API BufferedStreamReader /* not final */
+{
+public:
+   static constexpr size_t RequiredAlignment = 8;
+
+   explicit BufferedStreamReader(size_t bufferSize = 4096);
```

Comment:
Should you shrink the default to be sure the operating system doesn't grab a second 4K page that is mostly wasted?

Reply:
I'm not sure if I fully understand your concern. In case of a page fault, OS always will fetch more than one page from the swap. In case the memory read can be satisfied from RAM there is a balance of how many cache lines we are willing to waste for the buffer purposes vs the cost of reading data. 

4k is usually a good compromise, but the optimal performance for SQLite reader proved to be around 32k, otherwise, the cost of reading itself reduced performance noticeably. 

Speaking of how virtual memory is allocated the page size is rarely relevant too. And we have no control over it anyway, most likely an allocation will occupy the adjacent pages.

`fread` buffer size is 4k by default in Microsoft CRT, 8kb on GLIBC.

Please keep in mind, that in our configuration SQLite does not use memory-mapped IO, reading from disk usually is not implemented via memory mapping.

Reply:
I was thinking of allocating the memory for a vector of 4096 chars, which will have some small memory allocaiton overhead.  Does the memory allocator try to page-align such big requests, and if so, is there any small advantage in shrinking the vector size just a little so that data plus overhead fits that page?